lib/puppet/parser/scope.rb

```diff
@@ -615,7 +628,7 @@ def lookup_qualified_variable(class_name, variable_name, options)
           lookupvar(variable_name, options)
         end
       else
-        qualified_scope(class_name).lookupvar(variable_name, options)
+        qualified_scope(class_name).lookupvar(variable_name, options.merge({ :fqn_leaf => true }))
```

Comment:
The handling of `:fqn_leaf` seems a bit expensive. Potentially, there will be two extra hash lookups and then a merge, which in turn creates a new copy of the options hash.
Given that this code is very performance critical (I recently optimized it for that reason), I'd suggest either making `fqn_leaf` a third optional argument to the `#lookupvar` method, or to extract it into a variable once at the beginning of the method and then use that variable to avoid repeated hash lookups and also to make the merge optional.